From de9a85f249cb7dc167df419d64e38515996d4007 Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Thu, 22 Oct 2020 15:38:08 +0800
Subject: [PATCH] PONRTSYS-8097:net: EEE init on each link change event

Check EEE enabled or disabled on every link change event
Setup the EEE feature in MAC and PHY.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c | 53 ++++++++++++++++--------
 1 file changed, 36 insertions(+), 17 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
index a85b8e755e1a..d9857d724b2a 100644
--- a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
+++ b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
@@ -355,34 +355,47 @@ static int  ethtool_eee_get(struct net_device *dev,
 	return phy_ethtool_get_eee(dev->phydev, eee_data);
 }
 
-static int  ethtool_eee_set(struct net_device *dev,
-			    struct ethtool_eee *eee_data)
+static int  mac_set_eee(struct net_device *dev,
+			bool enable)
 {
-	struct ltq_eth_priv *priv = netdev_priv(dev);
+	int retval;
 	struct core_ops *ops = gsw_get_swcore_ops(0);
-	GSW_portLinkCfg_t	port_link_cfg = {0};
-	int retval = -EOPNOTSUPP;
+	GSW_portLinkCfg_t   port_link_cfg = {0};
 
 	if (!ops)
 		return -ENODEV;
+
 	port_link_cfg.nPortId = dev->dev_id;
 	retval = ops->gsw_common_ops.PortLinkCfgGet(ops, &port_link_cfg);
-
 	if (retval != 0) {
 		pr_err("%s: gsw PortLinkCfgGet failed\n", __func__);
 		return retval;
 	}
 
+	port_link_cfg.bLPI = enable;
+
+	retval = ops->gsw_common_ops.PortLinkCfgSet(ops, &port_link_cfg);
+	if (retval != 0) {
+		pr_err("%s: gsw PortLinkCfgSet failed\n", __func__);
+		return retval;
+	}
+
+	return 0;
+}
+
+static int  ethtool_eee_set(struct net_device *dev,
+			    struct ethtool_eee *eee_data)
+{
+	struct ltq_eth_priv *priv = netdev_priv(dev);
+	int retval = -EOPNOTSUPP;
+
 	priv->eee_enabled = eee_data->eee_enabled;
 	priv->tx_lpi_enabled = eee_data->tx_lpi_enabled;
 	priv->tx_lpi_timer = eee_data->tx_lpi_timer;
-	/* xgmac */
-	port_link_cfg.bLPI = eee_data->tx_lpi_enabled && eee_data->eee_enabled;
-
-	retval =  ops->gsw_common_ops.PortLinkCfgSet(ops, &port_link_cfg);
 
-	if (retval != 0) {
-		pr_err("%s: gsw PortLinkCfgset failed\n", __func__);
+	retval = mac_set_eee(dev, priv->eee_enabled && priv->tx_lpi_enabled);
+	if (retval) {
+		pr_err("%s: MAC LPI setting  failed\n", __func__);
 		return retval;
 	}
 
@@ -395,10 +408,6 @@ static int  ethtool_eee_set(struct net_device *dev,
 			ops->xgmac_reg_wr(ops, 0x00D4, priv->tx_lpi_timer);
 	}
 
-	/* phy */
-	if (eee_data->eee_enabled)
-		priv->eee_active = phy_init_eee(dev->phydev, 0) >= 0;
-
 	return retval;
 }
 
@@ -2361,7 +2370,7 @@ static void xrx500_gmac_update(struct xrx500_port *port)
 static void xrx500_mdio_link(struct net_device *dev)
 {
 	struct ltq_eth_priv *priv;
-	int i;
+	int i, retval;
 
 	priv = netdev_priv(dev);
 	pr_debug("%s called..\n", __func__);
@@ -2383,6 +2392,16 @@ static void xrx500_mdio_link(struct net_device *dev)
 					    dev->name);
 #endif
 			phy_print_status(priv->port[i].phydev);
+			if (priv->eee_enabled) {
+				priv->eee_active = !phy_init_eee(dev->phydev, 0);
+				retval = mac_set_eee(dev, priv->eee_enabled &&
+						     priv->tx_lpi_enabled);
+				if (retval) {
+					pr_err("%s:MAC LPI setting failed\n", __func__);
+					return;
+				}
+			}
+
 		}
 	}
 }
